lib/types.ts

```diff
@@ -0,0 +1,85 @@
+import type * as typesObject from './constant/types'
+import type {ValueOf} from 'type-fest'
+
+export interface Point {
+  line: number
+  column: number
+  offset: number
+  _index: number
+  _bufferIndex: number
+}
+export interface Token {
+  start: Point
+  end: Point
+  type: unknown
+  contentType: string
+  previous: Token
+  next: Token
+  _size: number
+  _break?: boolean
+  _tokenizer?: unknown
+  _events?: Event[]
+  _subevents?: Event[]
+  _contentTokenized?: unknown
+  _close?: unknown
+  _marker?: unknown
+  _open?: unknown
+  _side?: number
+  _tabs?: unknown
+}
+
+export type Type = ValueOf<typeof typesObject>
+
+export type Event = [name: string, token: Token, context: unknown]
+
+export interface Effects {
+  enter: (type: Type) => Token
+  consume: (code: number) => Token
+  exit: (type: Type) => Token
```

Comment:
Enter and exit define where tokens start and end. Consume deals with a character, and moves to the next.

Reply:
does consume not return a `Token`?

Reply:
nope, returns void.

Enter and exit do return token.